### PR TITLE
POC: Incremental Predicates

### PR DESCRIPTION
**Pull request by dm03514**
refs #3293

### Description

This PR updates the snowflake delete+insert incrementnal load stratgy to allow for the config to specify incremental predicates:

```
{{
    config(
      materialized='incremental',
      incremental_strategy='delete+insert',
      incremental_predicates=[
        "collector_hour >= dateadd('day', -7, CONVERT_TIMEZONE('UTC', current_timestamp()))"
      ],
      unique_key='unique_id'
    )
}}
```

I believe that this is distinct from the `predicates` which is already present in some of the functions. It looks like `predicates` applies directly to merge column. `incremental_predicates` are applied globally to the top level DML.

### TODO
- [x] default incremental - add param
- [x] default incremental upsert - add param 
- [x] common/merge - add param
- [ ] snowflake merge
- [x] snowflake incremental 
- [ ] bigquery incremental 
- [ ] bigquery merge
- [ ] postgres incremental
- [ ] postgres merge
- [ ] redshift incremental
- [ ] redshift merge

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [ ] This PR includes tests, or tests are not required/relevant for this PR
 - [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/include/global_project/macros/materializations/common/merge.sql [hunk 1]
```diff
@@ -1,25 +1,26 @@
 
 
-{% macro get_merge_sql(target, source, unique_key, dest_columns, predicates=none) -%}
-  {{ adapter.dispatch('get_merge_sql')(target, source, unique_key, dest_columns, predicates) }}
+{% macro get_merge_sql(target, source, unique_key, dest_columns, predicates=none, incrementnal_predicates=none) -%}
```
Comment (jtcohen6) [inline core/dbt/include/global_project/macros/materializations/common/merge.sql:3]:
```suggestion
{% macro get_merge_sql(target, source, unique_key, dest_columns, predicates=none, incremental_predicates=none) -%}
```

File: core/dbt/include/global_project/macros/materializations/common/merge.sql [hunk 2]
```diff
@@ -1,25 +1,26 @@
...
 
-{% macro get_insert_overwrite_merge_sql(target, source, dest_columns, predicates, include_sql_header=false) -%}
-  {{ adapter.dispatch('get_insert_overwrite_merge_sql')(target, source, dest_columns, predicates, include_sql_header) }}
+{% macro get_insert_overwrite_merge_sql(target, source, dest_columns, predicates, include_sql_header=false, incremental_predicates=none) -%}
+  {{ adapter.dispatch('get_insert_overwrite_merge_sql')(target, source, dest_columns, predicates, include_sql_header, incremental_predicates) }}
 {%- endmacro %}
 
 
-{% macro default__get_merge_sql(target, source, unique_key, dest_columns, predicates) -%}
+{% macro default__get_merge_sql(target, source, unique_key, dest_columns, predicates, incremental_predictates=none) -%}
```
Comment (jtcohen6) [inline core/dbt/include/global_project/macros/materializations/common/merge.sql:18]:
```suggestion
{% macro default__get_merge_sql(target, source, unique_key, dest_columns, predicates, incremental_predicates=none) -%}
```

**Review by jtcohen6 — COMMENTED** [review]
> I believe that this is distinct from the `predicates` which is already present in some of the functions. It looks like `predicates` applies directly to merge column. `incremental_predicates` are applied globally to the top level DML.

Is that so? I've definitely been thinking about these as the same thing: a way to filter the _target_ table, and limit the amount of data scanned. In `delete+insert` (upsert) operations, that just looks like a filter on the bottom, but in `merge` operations, I think that filter rightly goes on the join/match criteria. I may be thinking about this wrong, so let me know if you disagree.

File: core/dbt/include/global_project/macros/materializations/common/merge.sql (hunk 2 above)
Comment (dm03514) [reply core/dbt/include/global_project/macros/materializations/common/merge.sql:18]: @jtcohen6 sorry i def think you're right on re the predicates. We've adopted both strategies internally, and we are using the predicates directly on the merge statement, as you mentioned. I was thinking about it wrong :p

**Pull request closed without merging**